core/transaction-graph/src/future.rs

```diff
@@ -0,0 +1,157 @@
+	hash,
+};
+
+use sr_primitives::transaction_validity::{
+	TransactionTag as Tag,
+};
+
+use pool::Transaction;
+
+#[derive(Debug)]
+pub struct WaitingTransaction<Hash> {
+	pub transaction: Transaction<Hash>,
+	pub missing_tags: HashSet<Tag>,
+}
+
+impl<Hash> WaitingTransaction<Hash> {
+	pub fn new(transaction: Transaction<Hash>, provided: &HashMap<Tag, Hash>) -> Self {
+		let missing_tags = transaction.requires
+			.iter()
+			.filter(|tag| !provided.contains_key(&**tag))
+			.cloned()
+			.collect();
+
+		WaitingTransaction {
+			transaction,
+			missing_tags,
+		}
+	}
+
+	pub fn satisfy_tag(&mut self, tag: &Tag) {
+		self.missing_tags.remove(tag);
+	}
+
+	pub fn is_ready(&self) -> bool {
+		self.missing_tags.is_empty()
+	}
+}
+
+#[derive(Debug)]
+pub struct FutureTransactions<Hash: hash::Hash + Eq> {
+	/// tags that are not yet provided by any transaction and we await for them
+	wanted_tags: HashMap<Tag, HashSet<Hash>>,
+	/// Transactions waiting for a particular other transaction
+	waiting: HashMap<Hash, WaitingTransaction<Hash>>,
+}
+
+impl<Hash: hash::Hash + Eq> Default for FutureTransactions<Hash> {
+	fn default() -> Self {
+		FutureTransactions {
+			wanted_tags: Default::default(),
+			waiting: Default::default(),
+		}
+	}
+}
+
+impl<Hash: hash::Hash + Eq + Clone> FutureTransactions<Hash> {
+	/// Import transaction to Future queue.
+	///
+	/// Only transactions that don't have all their tags satisfied should occupy
+	/// the Future queue.
+	/// As soon as required tags are provided by some other transactions that are ready
+	/// we should remove the transactions from here and move them to the Ready queue.
+	pub fn import(&mut self, tx: WaitingTransaction<Hash>) {
+		assert!(!tx.is_ready(), "Transaction is ready.");
+		assert!(!self.waiting.contains_key(&tx.transaction.hash), "Transaction is already imported.");
+
+		// Add all tags that are missing
+		for tag in &tx.missing_tags {
+			let mut entry = self.wanted_tags.entry(tag.clone()).or_insert_with(HashSet::new);
+			entry.insert(tx.transaction.hash.clone());
+		}
+
+		// Add the transaction to a by-hash waiting map
+		self.waiting.insert(tx.transaction.hash.clone(), tx);
+	}
+
+	/// Returns true if given hash is part of the queue.
+	pub fn contains(&self, hash: &Hash) -> bool {
+		self.waiting.contains_key(hash)
+	}
+
+	/// Satisfies provided tags in transactions that are waiting for them.
+	///
+	/// Returns (and removes) transactions that became ready after their last tag got
+	/// satisfied and now we can remove them from Future and move to Ready queue.
+	pub fn satisfy_tags<T: AsRef<Tag>>(&mut self, tags: impl IntoIterator<Item=T>) -> Vec<WaitingTransaction<Hash>> {
+		let mut became_ready = vec![];
+
+		for tag in tags {
+			if let Some(hashes) = self.wanted_tags.remove(tag.as_ref()) {
+				for hash in hashes {
+					let is_ready = {
+						let mut tx = self.waiting.get_mut(&hash)
+							.expect("Every transaction in wanted_tags is present in waiting; qed");
```

Comment:
not strict enough. either prove how it is that they are present or manage the case where it's not gracefully.